Full support of special attributes in `migrate`

Diamondemon opened the pull request:
This pull request completes pull request #6126 by making the `gitattr` package able to access special attributes files. It paves the way for thorough correction of #5173, #5179 and #5181
New method for the Tree class:

- `FindSpecialAttributes` allows the Tree to have access to the system-level, user-level and repository-specific attributes files.

### Detailed behaviour:

- The gitattr.Tree structure uses its reference to the 3 special attributes trees, namely system, user and repository. These references can be filled by `Tree.FindSpecialAttributes` or stay `nil`.
- `FindSpecialAttributes` takes as arguments the config environments that contain information about where to find the system-level and user-level special attributes files, and the path of the git working directory to be able to get the repository-specific special attributes. It then builds their respective trees and keeps them into the respective references.
- These new trees are already taken into account by the `Applied` method.
- The new way to use the `Tree` class to its full potential is the following: `t = New --> t.FindSpecialAttributes --> t.Applied (as many times as necessary)` This order must be respected due to the mecanism in Applied that avoids re-processing macros.

### Other features:

- in order to be able to access special attributes files from the `gitattr` package without creating a dependency loop, the `attribs.go` file (which contains such logic) has been moved from the `git` package to the `git/gitattr` package.
- A new `git.Var` function has been added to be able to access git environment variables necessary for the retrieval of the aforementioned files (thanks @chrisd8088 ).
- In `attribs.go`, file-retrieval logic has been separated from the `AttributePath` parsing logic. `Tree.FindSpecialAttributes` uses the former only.
- Functions to get special attributes files that are used in `command_track.go` have been renamed to better fit their actual effect.

### Work left to do in later PRs:
- Make full use of this new logic in `command_track.go` and any other file that parses attributes, to have a unified process
- Full documentation of the `gitattr` package will be done once main related issues are resolved.

Diamondemon commented:
I fixed it. I did not understand the `defer` keyword the right way so I was using it wrong.

I also changed the first commit's message, changed the `attribs.go` file to `files.go`, and rearranged the number of commits.

Only missing point should be some new shell tests.

Diamondemon commented:
I looked into the shell tests and I realized that the system `gitattributes` file cannot really be checked in tests. There does not seem to be a really safe way to change it since the path `/etc/gitattributes` cannot be overridden the same way `$HOME` can be changed.
Thus we ignore it, right?

I created a test that at least verifies that both the user-specific attribute file and the repo-wide file are parsed and priorities are applied in the appropriate order.

I did not check for the override of macros since this is done both in the `macro_test.go` and `tree_test.go` files.

The test failed on Windows because of the user-level file I create. I there an equivalent on Windows?

chrisd8088 commented:
> The test failed on Windows because of the user-level file I create. I there an equivalent on Windows?

The global Git attributes file is definitely supported on Windows.  (As you know, Git's documentation uses "global" to mean "affects all of a user's repositories", not "affects all users' repositories", since that's what the system attributes file does.)

I looked at some of the CI jobs runs for this PR and I suspect this particular [run](https://github.com/git-lfs/git-lfs/actions/runs/20682472794/attempts/1) may be the one which suggested to you that the tests you'd written in your commit 41ca81e0a2458f99635d626a8d27 were failing on Windows due to a problem with the global attributes file.

However, I don't actually see a failure of the `t-migrate*.sh` scripts in that job; rather, I think we had one of our (unfortunately not uncommon) test flakes on Windows, in this case in the unrelated `t-status.sh` test script.

I re-ran that particular CI job and while there were still [failures](https://github.com/git-lfs/git-lfs/actions/runs/20682472794/attempts/2), they are definitely unrelated to your PR.  The tests in the `t-migrate*.sh` scripts all passed, as did those in the `t-status.sh` script.  The "new" failures are due to a known regression in upstream Git which made it way into Git for Windows 2.53.0.rc0.windows.1, and appear in the `t-fetch-include.sh`, `t-unusual-filenames.sh`, and `t-zero-len-file.sh` scripts.  Once Git 2.53.0-rc1 is tagged and a corresponding Git for Windows tag is made, we should see these failures disappear from our CI suite.

In the meantime, though, I don't think you need to exempt your tests of the global attributes file from running on Windows, so you probably can drop your commit 554c569aef9d485e6cbc6bceeb8459121d5f622a.

---

One detail which may be worth stating with respect to our Windows tests is that we run them in the Git Bash environment, which uses MSYS2 to emulate aspects of a Unix system.  As a consequence, if you construct a path like `$HOME/.git/config/attributes` using the `HOME` environment variable, it will look something like `/tmp/git-lfs_TEMP.lITl9S/t-migrate-fixup.sh-1440/home/.git/config/attributes`, with a leading `/tmp` directory.  But if you ask Git to report the path to the global attributes file with `git var GIT_ATTR_GLOBAL`, it will look like `C:/Users/RUNNER~1/AppData/Local/Temp/git-lfs_TEMP.lITl9S/t-migrate-fixup.sh-1440/home/.config/git/attributes`.  Both of these paths refer to the same filesystem location, but the first will be translated by MSYS2, while the latter is the "canonical" Windows path.

---

If you want to be explicit, you could set the path to the global attributes file by setting the `core.attributesFile` configuration variable, instead of creating a file in `$HOME/.git/config/attributes`.  For example, you could do:

```
global="$(canonical_path "$TRASHDIR/gitattributes-global")"
git config core.attributesFile "$global"

echo "*.bat filter=overridden diff=lfs merge=lfs -text" >"$global"
```

That should function on all systems too, while using an explicitly specified path.  (Note also that it's OK to set global Git configuration values because we reset the global configuration file at the start of each test.)

chrisd8088 commented on git/gitattr/files.go:
```diff
@@ -47,44 +47,86 @@ func (s *AttributeSource) String() string {
-	af, _ := cfg.Get("core.attributesfile")
-	af, err := tools.ExpandConfigPath(af, "git/attributes")
+// GetRepoAttributePaths behaves as GetAttributePaths, and loads information
+// only from the repo attributes file.
```
I think this comment should read:

```suggestion
// GetUserAttributePaths behaves as GetAttributePaths, and loads information
// only from the user's global attributes file.
```

Diamondemon commented on git/gitattr/tree.go:
```diff
@@ -88,6 +88,40 @@ func NewFromReader(mp *MacroProcessor, rdr io.Reader) (*Tree, error) {
+	}
+	if systemReader != nil {
+		t.systemAttributes, err = NewFromReader(t.mp, systemReader)
+	}
```
I realized I did understand the `defer` keyword in the wrong way then changed something in `files.go` and I forgot to edit the `tree.go` file.

Diamondemon commented:
I read all your comments so here are the differences in my new push:

- I removed the Windows condition. Somehow I had not read the test result properly and blamed my code.
- I changed the documentation addition of `setup_single_local_branch_tracked_corrupt` to the phrasing you suggested.
- I fixed the typo you mentioned on `overridden`
- I added the quotes around `$HOME` in my tests
- I added reworked the file opening logic as per your suggestion
- I renamed all the methods to `Get*AttributeFilePath` for consistency
- I put my unrelated warning fix out of the PR for now.

> One detail which may be worth stating with respect to our Windows tests is that we run them in the Git Bash environment, which uses MSYS2 to emulate aspects of a Unix system. As a consequence, if you construct a path like $HOME/.git/config/attributes using the HOME environment variable, it will look something like /tmp/git-lfs_TEMP.lITl9S/t-migrate-fixup.sh-1440/home/.git/config/attributes, with a leading /tmp directory. But if you ask Git to report the path to the global attributes file with git var GIT_ATTR_GLOBAL, it will look like C:/Users/RUNNER~1/AppData/Local/Temp/git-lfs_TEMP.lITl9S/t-migrate-fixup.sh-1440/home/.config/git/attributes. Both of these paths refer to the same filesystem location, but the first will be translated by MSYS2, while the latter is the "canonical" Windows path. 
> [...]

Does that mean that one way should be better than the other?

> As for system-level attribute files, let's not complicate this PR further by trying to test those, although as I described in reply to your question, it's not completely out of the question. But we'd have to revise some aspects of our GitHub Actions workflows to make this possible, and while I think that would be a good idea, I also think it expands the scope of this PR beyond what's reasonable.

Yeah, that goes a bit out of scope indeed.

> Thank you so much for this PR, it was a pleasure to read! Each commit has a nicely isolated and logical set of changes, which makes it so much easier to understand how they build on each other.

Thanks, I appreciate that. I'm glad this works well with you.

chrisd8088 commented on commands/command_merge_driver.go:
```diff
@@ -132,7 +132,7 @@ func mergeProcessInput(gf *lfs.GitFilter, filename string, fileSpecifiers map[st
 	}
 	defer fp.Close()
-	_, err = gf.Smudge(file, pointer, file.Name(), true, getTransferManifestOperationRemote("download", cfg.Remote()), cb)
+	_, _ = gf.Smudge(file, pointer, file.Name(), true, getTransferManifestOperationRemote("download", cfg.Remote()), cb)
```
> I think that's just the official Go extension in Visual Studio Code. 

Interesting, thank you!  And thanks for moving this change out of the PR!
